### PR TITLE
add unary quadratic_map for SPZ

### PR DESCRIPTION
**Pull request by lucaferranti**

### CODE REVIEW

File: src/Approximations/overapproximate.jl [hunk 1]
```diff
@@ -1963,6 +1963,43 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+"""
+    overapproximate(overapproximate(QM::QuadraticMap, ::Type{SparsePolynomialZonotope}))
+
+Return a sparse polynomial zonotope overapproximation of the quadratic map of a sparse
+polynomial zonotope. Implements Proposition 3.1.31 of [1].
+
+[1] N. Kochdumper. *Extensions of polynomial zonotopes and their application to
+verification of cyber-physical systems*. 2021.
+"""
+function overapproximate(QM::QuadraticMap, ::Type{SparsePolynomialZonotope})
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1975]: You need to constrain the `QuadraticMap`'s set type (otherwise this would be incorrect).

```suggestion
"""
    overapproximate(QM::QuadraticMap{N, <:SparsePolynomialZonotope},
                    ::Type{SparsePolynomialZonotope})

Return a sparse polynomial zonotope overapproximation of the quadratic map of a
sparse polynomial zonotope.

### Input

- `QM` -- quadratic map of a sparse polynomial zonotope
- `SparsePolynomialZonotope` -- target type

### Output

A sparse polynomial zonotope overapproximating the quadratic map of a sparse
polynomial zonotope.

### Algorithm

This method implements Proposition 3.1.31 of [1].

[1] N. Kochdumper. *Extensions of polynomial zonotopes and their application to
verification of cyber-physical systems*. 2021.
"""
function overapproximate(QM::QuadraticMap{N, <:SparsePolynomialZonotope},
                         ::Type{SparsePolynomialZonotope})
```

File: src/Approximations/overapproximate.jl [hunk 2]
```diff
@@ -1963,6 +1963,43 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+	PZ = QM.X
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1976]: There are again mixed tabs and spaces.

File: src/Sets/SparsePolynomialZonotope.jl [hunk 3]
```diff
@@ -350,31 +350,3 @@ function remove_redundant_generators(S::SparsePolynomialZonotope)
     GI = remove_zero_columns(genmat_indep(S))
     return SparsePolynomialZonotope(c, G, GI, E)
 end
-
-# function quadratic_map(Q::Vector{MT}, S::SparsePolynomialZonotope) where {N, MT<:AbstractMatrix{N}}
```
Comment (schillic) [inline src/Sets/SparsePolynomialZonotope.jl:354]: Why did you remove this method?

File: test/Sets/SparsePolynomialZonotope.jl [hunk 4]
```diff
@@ -77,3 +77,17 @@ for Z in [rand(Zonotope), rand(Hyperrectangle)]
...
+    PZ1 = SparsePolynomialZonotope([0.0, 0.0], [1. -1 1;-1 2 1], hcat([0.1, 0]), [1 0 2;0 1 1], 1:2)
+    Q = QuadraticMap([[0.5 0.5;1 -0.5], [-1.0 0;1 0]], PZ1)
+    QPZ = overapproximate(Q, SparsePolynomialZonotope)
+    @test center(QPZ) ≈ [0.0025, -0.005]
+    @test genmat_dep(QPZ) ==  [-1.5  5.5   2.0  -4.5  -1.5  1.5;
+                               -2.0  5.0  -2.0  -3.0   3.0  0.0]
+    @test genmat_indep(QPZ) ≈ [-0.05 0.15 0.15 0 0.05 0.1 0.0025;
+                                       -0.2  0.3  0  -0.1 0.1 -0.1 -0.005]
+    @test expmat(QPZ) ==  [2  1  3  0  2  4;
+                                  0  1  1  2  2  2]
```
Comment (schillic) [inline test/Sets/SparsePolynomialZonotope.jl:92]:
```suggestion
    @test genmat_indep(QPZ) ≈ [-0.05 0.15 0.15 0 0.05 0.1 0.0025;
                               -0.2  0.3  0  -0.1 0.1 -0.1 -0.005]
    @test expmat(QPZ) == [2  1  3  0  2  4;
                          0  1  1  2  2  2]
```

**Review by schillic — COMMENTED** [review]
Rather use Proposition 13 in [1], which is specialized for the same SPZ.

- [KA21] N. Kochdumper and M. Althoff. *Sparse Polynomial Zonotopes: A Novel Set Representation for Reachability Analysis*. Transactions on Automatic Control, 2021.

File: src/Sets/SparsePolynomialZonotope.jl (hunk 3 above)
Comment (lucaferranti) [reply src/Sets/SparsePolynomialZonotope.jl:354]: this was a copy-paste of the code for SSPZ (note it was commented out), meant to be a starting point for the implementation for SPZ. Since the method computes an overapproximation, I understood the right way is to define `overapproximate(::QuadraticMap, ::Type{SparsePolynomialZonotope})` as I did in this PR. Should I also have a `quadratic_map(Q, S::SparsePolynomialZonotope)` ?

Comment (schillic) [reply src/Sets/SparsePolynomialZonotope.jl:354]: Ah okay, I did not see that it was commented out. Then it is fine.
No, we do not want to use `quadratic_map` for overapproximations.

**Comment by lucaferranti** [comment]
interestingly, using the version for the paper (specialized for 1 set) returns a different result with less independent generators... I wonder if it's the same but just optimized taking into account that it's the same set

File: src/Approximations/overapproximate.jl [hunk 5]
```diff
@@ -1963,6 +1963,42 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
     return Zonotope(d, remove_zero_columns(h))
 end
 
+"""
+    overapproximate(overapproximate(QM::QuadraticMap, ::Type{SparsePolynomialZonotope}))
+
+Return a sparse polynomial zonotope overapproximation of the quadratic map of a sparse
+polynomial zonotope. Implements Proposition 13 of [1].
+
+[1] N. Kochdumper and M. Althoff. Sparse Polynomial Zonotopes: A Novel Set Representation for Reachability Analysis. Transactions on Automatic Control, 2021.
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1972]:
```suggestion
    overapproximate(QM::QuadraticMap{N, <:SparsePolynomialZonotope},
                    ::Type{SparsePolynomialZonotope}) where {N}

Return a sparse polynomial zonotope overapproximating the quadratic map of a
sparse polynomial zonotope.

### Input

- `QM` -- quadratic map of a sparse polynomial zonotope
- `SparsePolynomialZonotope` -- target type

### Output

A sparse polynomial zonotope overapproximating the quadratic map of a sparse
polynomial zonotope.

### Algorithm

This method implements Proposition 3.1.31 of [1].

[1] N. Kochdumper. *Extensions of polynomial zonotopes and their application to
verification of cyber-physical systems*. 2021.
```

File: src/Approximations/overapproximate.jl [hunk 6]
```diff
@@ -1963,6 +1963,58 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+- `SparsePolynomialZonotope` -- target type
+
+### Output
+
+A sparse polynomial zonotope overapproximating the quadratic map of a sparse
+polynomial zonotope.
+
+### Algorithm
+
+This method implements Proposition 3.1.31 of [1].
```
Comment (lucaferranti) [inline src/Approximations/overapproximate.jl:1985]:
```suggestion
This method implements Proposition 13 of [1].
```

File: src/Approximations/overapproximate.jl [hunk 7]
```diff
@@ -1963,6 +1963,58 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+
+[1] N. Kochdumper. *Extensions of polynomial zonotopes and their application to
```
Comment (lucaferranti) [inline src/Approximations/overapproximate.jl:1987]:
```suggestion
[1] N. Kochdumper and M. Althoff. Sparse Polynomial Zonotopes: A Novel Set Representation for
```

File: src/Approximations/overapproximate.jl [hunk 8]
```diff
@@ -1963,6 +1963,58 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+verification of cyber-physical systems*. 2021.
```
Comment (lucaferranti) [inline src/Approximations/overapproximate.jl:1988]:
```suggestion
Reachability Analysis. Transactions on Automatic Control, 2021.
```

File: src/Approximations/overapproximate.jl [hunk 9]
```diff
@@ -1963,6 +1963,58 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+    Ê = cat(expmat(PZ), Matrix(1 * I, q, q); dims=(1, 2))
+
+    PZ1 = SimpleSparsePolynomialZonotope(c, Ĝ, Ê)
+    PZbar = quadratic_map(Q, PZ1)
+    c̄ = center(PZbar)
+    Ē = expmat(PZbar)
+    Ḡ = genmat(PZbar)
+
+    H = [iszero(Ei[p+1:end]) for Ei in eachcol(Ē)]
+    K = .!H
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:2010]: I suggest to reorder the computation to stay closer to the source.
```suggestion
    K = [!iszero(Ei[p+1:end]) for Ei in eachcol(Ē)]
    H = .!K
```

File: src/Approximations/overapproximate.jl [hunk 10]
```diff
@@ -1963,6 +1963,58 @@ function overapproximate(qm::QuadraticMap{N, <:AbstractZonotope}, ::Type{<:Zonot
...
+    Q = QM.Q
+
+    p = nparams(PZ)
+    q = ngens_indep(PZ)
+
+    c = center(PZ)
+    G = genmat_dep(PZ)
+    GI = genmat_indep(PZ)
+    Ĝ = hcat(G, GI)
+    Ê = cat(expmat(PZ), Matrix(1 * I, q, q); dims=(1, 2))
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:2001]: At some point we have to review these constructions to make use of sparsity. But we can keep it for now.

**Review by schillic — APPROVED** [review]

File: src/Approximations/overapproximate.jl (hunk 10 above)
Comment (lucaferranti) [reply src/Approximations/overapproximate.jl:2001]: https://github.com/JuliaReach/LazySets.jl/issues/3068 opened an issue for it
